# Data to be published
experiment         : 1 # Experiment A=0, B=1, C=2

# Flexiforce settings
port               : /dev/ttyACM0
baud               : 2000000
tactile_frame_id   : l_gripper_tool_frame
num_tactile_sensors: 4
sensor_0 : [-1,0,0] # Opposite of sensor normal
sensor_1 : [0,1,0]
sensor_2 : [1,0,0]
sensor_3 : [0,-1,0]
max_calibration_distance: 0.2

loopRateFactor     : 3 # Virtual loop rate [ms] (time for computations = loopRateFactor - 2)

useCurrentCartPose : false
useNullspacePose   : true
useFlexiForce      : false
forceTorqueOn      : false
accelerometerOn    : false
useDigitalFilter   : false

executeCircleTraj  : false
numCircleTraj      : 3

mannequinMode      : true
mannequinThresPos  : 0.05
mannequinThresRot  : 0.05 # not used

publishRTtopics    : true

# Specifically for MRAC
useSimHuman        : false

useOuterloop       : false
# Only one can be true
useARMAmodel       : true
useCTARMAmodel     : false
useFIRmodel        : false
useMRACmodel       : false
useIRLmodel        : false
useMSDmodel        : false
useDirectmodel     : false

outerLoop_time     : 0.0

#### Digital filter coefficients
## Lowpass, 1st order butterworth
# fc=1000Hz, fs=3000Hz
#a_filt: [1.0, 0.2679]
#b_filt: [0.634, 0.634]
# fc=35Hz, fs=1000Hz
a_filt: [1.0, -0.8012]
b_filt: [0.0994, 0.0994]
## Band-pass, 1st order chebychev
# 5-1000Hz, fs=3000Hz
#a_filt: [1.0,-0.3329,-0.6610]
#b_filt: [0.8305, 0, -0.8305]
# 5-100Hz, fs=1000Hz
#a_filt: [1.0,-1.0526,0.0634]
#b_filt: [0.4683, 0, -0.4683]

# Filter for qdot values
joint_vel_filter : 1.0

# Is the reference trajectory x_r or q_r set externally or estimated
# If externalRefTraj is false the human intent x_i is used
# i.e. an admittance model is used to generate x_i from f_h the force from the FT sensor
externalRefTraj    : true

# Human intent estimation
useHumanIntent     : false
intentEst_time     : 0.05
intentEst_delT     : 0.1  # Propagation time
intentEst_M        : 100.0  # M_i the mass of the intent admittance model

# The params of the RLS estimator
rls_lambda         : 0.98
rls_sigma          : 0.001

# IRL
numIrlSamples      : 200
numIrlLsIter       : 20
numCartDof         : 1
irlOneshot         : true

# Params for the MRAC outer-loop controller
mrac_gamma_1       : 0.1
mrac_gamma_2       : 2.0
mrac_gamma_3       : 2.5
mrac_gamma_4       : 4
mrac_gamma_5       : 4

mrac_P_m           : 1
mrac_P_h           : 1

fixedFilterWeights : false

# ARMA parameters
filtW0                : -1.8775  #-1.8136 #-2.3646
filtW1                :  1.2729  # 1.3171 # 1.9237
filtW2                : -0.0578  #-0.4726 #-0.3255
filtW3                : -0.3269  #-0.0082 #-0.1816
filtW4                :  0.0018  # 0.0032 # 0.0049
filtW5                :  0.0001  #-0.0010 #-0.0018
filtW6                : -0.0013  #-0.0007 #-0.0018
filtW7                :  0.0002  #-0.0009 # 0.0014

# ARMA parameters for flexiforce          
flex_1_filtW0         :  0.0
flex_1_filtW1         :  0.0
flex_1_filtW2         :  0.0
flex_1_filtW3         :  0.0
flex_1_filtW4         :  0.0
flex_1_filtW5         :  0.0
flex_1_filtW6         :  0.0
flex_1_filtW7         :  0.0
                        
flex_2_filtW0         : -0.6283
flex_2_filtW1         : -0.3757
flex_2_filtW2         :  0.1192
flex_2_filtW3         : -0.1512
flex_2_filtW4         :  0.0063
flex_2_filtW5         : -0.0017
flex_2_filtW6         : -0.0030
flex_2_filtW7         :  0.0026
                      
flex_3_filtW0         :  0.0
flex_3_filtW1         :  0.0
flex_3_filtW2         :  0.0
flex_3_filtW3         :  0.0
flex_3_filtW4         :  0.0
flex_3_filtW5         :  0.0
flex_3_filtW6         :  0.0
flex_3_filtW7         :  0.0
                        
flex_4_filtW0         : -0.7185
flex_4_filtW1         :  0.1265
flex_4_filtW2         : -0.1395
flex_4_filtW3         : -0.2230
flex_4_filtW4         : -0.0009
flex_4_filtW5         :  0.0003
flex_4_filtW6         : -0.0001
flex_4_filtW7         :  0.0013

# This makes zero any force less than below
forceCutOffX          : 0.1
forceCutOffY          : 0.1
forceCutOffZ          : 0.1

# Params for the MSD outer-loop controller
m_M                   : 1.0
m_S                   : 0.0
m_D                   : 5.0

# Params for M(s) of the RLS models
task_mA               : 0.7
task_mB               : 0.7

# Params for MRAC Sim Human
simHuman_a            : 0.1
simHuman_b            : 2.6

#cartIniX              :  0.7
#cartIniY              :  0.15
#cartIniZ              : -0.1
#cartIniRoll           :  0.0
#cartIniPitch          :  0.0
#cartIniYaw            :  0.0

cartDesX              :  0.7
cartDesY              :  0.3
cartDesZ              : -0.1
cartDesRoll           :  0.0
cartDesPitch          :  0.0
cartDesYaw            :  0.0

cartDes2X             :  0.9
cartDes2Y             :  0.3
cartDes2Z             : -0.1
cartDes2Roll          :  0.0
cartDes2Pitch         :  0.0
cartDes2Yaw           :  0.0

cartDes3X             :  0.9
cartDes3Y             :  0.0
cartDes3Z             : -0.1
cartDes3Roll          :  0.0
cartDes3Pitch         :  0.0
cartDes3Yaw           :  0.0

## Initial Position
# lCartIniX           :  0.8
# lCartIniY           :  0.05
# lCartIniZ           :  0.0
# lCartIniRoll        :  1.57079633
# lCartIniPitch       :  0.0
# lCartIniYaw         : -1.57079633

cartIniX             :  0.65
cartIniY             :  0.35
cartIniZ             :  -0.10

#cartIniX             :  0.75
#cartIniY             :  0.35
#cartIniZ             :  0.0
cartIniRoll          :  1.57
cartIniPitch         :  0.0
cartIniYaw           :  0.0

circleRate            : 0.5 # angular velocity in rad/sec (v = r*w = A*w)
circleLlim            : 0.0 # amplitude = (circleUlim - circleLlim)/2
circleUlim            : 0.8 #  = A/2

nnNum_Inputs          : 44 
nnNum_Outputs         : 6  # Cartesian
nnNum_Hidden          : 12
nnNum_Error           : 6  
nnNum_Joints          : 7  # DOF

# Filtered error
# r = (qd_m - qd) + lambda*(q_m - q);
# Kv*r

nn_Kv                 : 5 # not being used? see cartPos_Kd and cartPos_Kp insteads
nn_lambda             : 20 # not being used?
nn_kappa              : 0.25
nn_Kz                 : 0.005 # not being used?
nn_Zb                 : 100 # not being used?
nn_feedForwardForce   : 1.0 #25.0 # small because of jitter from FT measurments
nn_nnF                : 100 # -> W (outer)
nn_nnG                : 200 # -> V (inner)
nn_ON                 : 1

# nn_Kv               : 5
# nn_lambda           : 20
# nn_kappa            : 0.07
# nn_Kz               : 0.001 
# nn_Zb               : 1000 
# nn_feedForwardForce : 0   
# nn_nnF              : 100 
# nn_nnG              : 100
# nn_ON               : 1

# Kd*(qd_m - qd) + Kp*(q_m - q)
cartPos_Kp_x          : 60
cartPos_Kp_y          : 60
cartPos_Kp_z          : 60
cartRot_Kp_x          : 10
cartRot_Kp_y          : 10
cartRot_Kp_z          : 10

# Prevents motion if qd=0
cartPos_Kd_x          : 10
cartPos_Kd_y          : 10
cartPos_Kd_z          : 10
cartRot_Kd_x          : 3
cartRot_Kd_y          : 3
cartRot_Kd_z          : 3

# Nominal angles for the arm joints
                # Left arm          # Right arm
nom_joint_0  :  0.0    # 0.067      # -0.48577  
nom_joint_1  :  0.0    # 0.036      # -0.0190721
nom_joint_2  :  1.5828 # 0.012      # -1.51115  
nom_joint_3  : -0.5    #-0.097      # -1.70928  
nom_joint_4  :  0.0    #-0.983      #  1.54561  
nom_joint_5  :  0.0    #-0.0947     #  0.046854 
nom_joint_6  :  0.0    # 0.1114     # -0.0436174


#gripper_acc_tip: r_gripper_motor_accelerometer_link
gripper_acc_tip: l_gripper_motor_accelerometer_link
ft_frame_id: l_force_torque_link #l_gripper_tool_frame

pr2_cartneuroController:
  type: ice_robot_controllers/PR2CartneuroControllerClass
  root_name: torso_lift_link
  tip_name: l_gripper_tool_frame

pr2_adaptNeuroController:
  type: ice_robot_controllers/PR2adaptNeuroControllerClass
  root_name: torso_lift_link
  tip_name: l_gripper_tool_frame

# Torque saturation
saturation:
  l_shoulder_pan_joint: 11.88
  l_shoulder_lift_joint: 11.64
  l_upper_arm_roll_joint: 6.143
  l_elbow_flex_joint: 6.804
  l_forearm_roll_joint: 8.376
  l_wrist_flex_joint: 5.568
  l_wrist_roll_joint: 5.568
  r_shoulder_pan_joint: 11.88
  r_shoulder_lift_joint: 11.64
  r_upper_arm_roll_joint: 6.143
  r_elbow_flex_joint: 6.804
  r_forearm_roll_joint: 8.376
  r_wrist_flex_joint: 5.568
  r_wrist_roll_joint: 5.568

# Posture control
k_posture: 500.0
jacobian_inverse_damping: 0.01
joint_feedforward:
  l_shoulder_pan_joint: 3.33
  l_shoulder_lift_joint: 1.16
  l_upper_arm_roll_joint: 0.1
  l_elbow_flex_joint: 0.25
  l_forearm_roll_joint: 0.133
  l_wrist_flex_joint: 0.0727
  l_wrist_roll_joint: 0.0727
  r_shoulder_pan_joint: 3.33
  r_shoulder_lift_joint: 1.16
  r_upper_arm_roll_joint: 0.1
  r_elbow_flex_joint: 0.25
  r_forearm_roll_joint: 0.133
  r_wrist_flex_joint: 0.0727
  r_wrist_roll_joint: 0.0727
#[0.467279,0.38766,0.850325,-1.88652,3.09594,-1.28368,-3.89919];
posture:
  l_shoulder_pan_joint: 0.4
  l_shoulder_lift_joint: 0.4
  l_upper_arm_roll_joint: 0.8
  l_elbow_flex_joint: -1.8
  l_forearm_roll_joint: 3.14
  l_wrist_flex_joint: -1.3
  l_wrist_roll_joint: -3.9
  r_shoulder_pan_joint: -0.5
  r_shoulder_lift_joint: 0.0
  r_upper_arm_roll_joint: -1.5
  r_elbow_flex_joint: -1.7
  r_forearm_roll_joint: 1.50
  r_wrist_flex_joint: 0.0
  r_wrist_roll_joint: 0.0
